lib/SymbolGraphGen/SymbolGraph.cpp

```diff
@@ -241,12 +238,31 @@ void SymbolGraph::recordMemberRelationship(Symbol S) {
       if (isRequirementOrDefaultImplementation(S.getSymbolDecl())) {
         return;
       }
+
       if (DC->getSelfNominalTypeDecl() == nullptr) {
         // If we couldn't look up the type the member is declared on (e.g.
         // because the member is declared in an extension whose extended type
         // doesn't exist), don't record a memberOf relationship.
         return;
       }
+
+      // If this is an extension to an external type, we use the extension
+      // symbol itself as the target.
+      if (auto const *Extension =
+              dyn_cast_or_null<ExtensionDecl>(DC->getAsDecl())) {
+        auto const shouldBeRecordedAsExtension =
+            AssociateMembersWithExtensionBlock &&
+            !Extension->getModuleContext()->getNameStr().equals(
```

Comment:
If I understood the documentation of your PR correctly, then the problem looked as follows:

Assuming we have a Swift module `A` with an underlying Clang module, then the Swift compiler handles the underlying Clang module by wrapping it in a Swift module also called `A`, which is then included in the Swift module `A` using an `@_exported import A` statement.

This then caused problems in SymbolGraphGen because the (C) `A` was known to be `@_exported` imported and the module comparison was exclusively done via the name, so everything in (Swift) `A` was also treated as if it were `@_exported` imported. Thus, extensions to external types such as `public extension String {}` were also exported to the MainGraph.

Is that about right?

Now, this is how I think that problem translates to my case:

The only case where the name equality comparison designates two different modules as equal is if it compares the (Swift) `A` with the (C) `A`. In my logic, this would be the case if (Swift) `A` extended a type in (C) `A`. This is possible. In that case, the string equality would consider the Swift/C modules equal and thus `shouldBeRecordedAsExtension` is set to `false`. This is consistent with the fact that the contents of this extension are stored in `A.symbols.json`, not `A@A.symbols.json` (which does not exist).

Is this the case you were concerned about? Does the behavior described above match what you expected to happen?

I tried to capture this behavior in [test/SymbolGraph/ClangImporter/ClangExtensions.swift](https://github.com/apple/swift/blob/89667d43306780922702ff5ce10331722177d179/test/SymbolGraph/ClangImporter/ClangExtensions.swift).